tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/ClientEncryptionProseTests.cs

```diff
@@ -0,0 +1,827 @@
+            {
+                Dictionary<string, BsonDocument> schemaMap = null;
+                if (schemaMapDocument != null)
+                {
+                    var element = schemaMapDocument.Single();
+                    schemaMap = new Dictionary<string, BsonDocument>
+                    {
+                        { element.Name, element.Value.AsBsonDocument }
+                    };
+                }
+
+                if (kmsProviders == null)
+                {
+                    kmsProviders = new ReadOnlyDictionary<string, IReadOnlyDictionary<string, object>>(new Dictionary<string, IReadOnlyDictionary<string, object>>());
+                }
+
+                var autoEncryptionOptions = new AutoEncryptionOptions(
+                    keyVaultNamespace,
+                    kmsProviders,
+                    schemaMap: Optional.Create<IReadOnlyDictionary<string, BsonDocument>>(schemaMap),
+                    extraOptions: extraOptions);
+
+                if (withExternalKeyVault)
+                {
+                    var externalKeyVaultClient = 
+                        new MongoClient(
+                            new MongoClientSettings
+                            {
+                                Credential = MongoCredential.FromComponents(null, null, "fake-user", "fake-pwd")
+                            });
+                    autoEncryptionOptions = autoEncryptionOptions
+                        .With(
+                            keyVaultClient: Optional.Create<IMongoClient>(externalKeyVaultClient));
+                }
+
+                mongoClientSettings.AutoEncryptionOptions = autoEncryptionOptions;
+            }
+
+            return new MongoClient(mongoClientSettings);
+        }
+
+        private void Insert(
+            IMongoCollection<BsonDocument> collection,
+            bool async,
+            params BsonDocument[] documents)
+        {
+            if (async)
+            {
+                collection
+                    .InsertManyAsync(documents)
+                    .GetAwaiter()
+                    .GetResult();
+            }
+            else
+            {
+                collection.InsertMany(documents);
+            }
+        }
+
+        public class JsonTestDataFactory : EmbeddedResourceJsonFileReader
+        {
+            #region static
+            private static JsonTestDataFactory __instance;
+            public static JsonTestDataFactory Instance => __instance ?? (__instance = new JsonTestDataFactory());
+            private static readonly string[] __ignoreKeyNames =
+            {
+                "dbPointer" // not supported
+            };
+            #endregion
+
+            private readonly IReadOnlyDictionary<string, BsonDocument> _documents;
+
+            public JsonTestDataFactory()
+            {
+                _documents = new ReadOnlyDictionary<string, BsonDocument>(ReadDocuments());
```

Comment:
It's a bit odd to be doing IO in a constructor but I guess this is OK since this is test code only.

Reply:
Note: the reason we wouldn't normally do IO in a constructor is that as a general rule we don't want constructors to throw exceptions (other than for invalid arguments) and IO can throw. We also as a general rule expect constructors to be fast and to next to no work and IO can be slow.